### PR TITLE
Remove ROCm and QTHREADS from the build system

### PR DESCRIPTION
**Pull request by masterleinad**
This pull request removes the `QTHREADS` and the `ROCm` backend from the build system, but I didn't touch `Makefile.kokkos` or `Makefile.targets`.
This way the backends should not be visible to users anymore.

### CODE REVIEW

**Comment by jeffmiles63** [comment]
see #2472 for the other 5 files (doc and scripts)

**Review by dalg24 — DISMISSED** [review]
Looks reasonable but can't approve before the confusion with #2472 and #2473 is resolved

**Comment by masterleinad** [comment]
I am happy to rebase this (if required) once #2473 is merged.

**Comment by masterleinad** [comment]
@dalg24 I see no confusion with the three pull requests. They are addressing separate concerns, don't conflict and remove the two backends from user visibility together. This might be the most controversial one so we can do it last.

@jeffmiles63 Do you agree?

File: generate_makefile.bash
Comment (dalg24) [inline generate_makefile.bash]: This change was not in #2472 but all other changes to that file were.

File: cmake/kokkos_arch.cmake
Comment (dalg24) [inline cmake/kokkos_arch.cmake]: We need to comment on why we leave empty if-statements here.

**Review by dalg24 — COMMENTED** [review]
> @dalg24 I see no confusion with the three pull requests. They are addressing separate concerns, don't conflict and remove the two backends from user visibility together. This might be the most controversial one so we can do it last.
> 
> @jeffmiles63 Do you agree?

Both of you are editing `scripts/testing_scripts/generate_makefile.bash`

@Rombur Do you reckon we should keep `KOKKOS_ARCH_ROCM`?

**Comment by masterleinad** [comment]
> Both of you are editing `scripts/testing_scripts/generate_makefile.bash`

I don't touch `scripts/testing_scripts/generate_makefile.bash` here, but only `generate_makefile.bash`.

**Comment by dalg24** [comment]
> > Both of you are editing `scripts/testing_scripts/generate_makefile.bash`
> 
> I don't touch `scripts/testing_scripts/generate_makefile.bash` here, but only `generate_makefile.bash`.

Sigh.  I forgot #2339 duplicated the file...

File: cmake/kokkos_arch.cmake
Comment (masterleinad) [reply cmake/kokkos_arch.cmake]: Do we want to keep the if-statements at all? I couldn't quite convince myself if it's better to remove them or keep them empty.

Removing these architectures completely might break someone's code even if we are not using it at all.

**Comment by dalg24** [comment]
Waiting to hear from Bruno to decide what to do with `KOKKOS_ARCH_ROCM`.  Everything else looks good.

File: cmake/kokkos_arch.cmake
Comment (Rombur) [reply cmake/kokkos_arch.cmake]: I don't understand what's the logic to keep this?

Comment (masterleinad) [reply cmake/kokkos_arch.cmake]: The only reason I could come up with is 'A placeholder for future flags, so we haven't "forgotten" about considering these architectures'.

Comment (Rombur) [reply cmake/kokkos_arch.cmake]: I don't think you can run HIP on hardware older than FIJI and I don't think we want to support anything older than Vega

Comment (masterleinad) [reply cmake/kokkos_arch.cmake]: So you are proposing that we remove all the architectures where we have empty flags (including `RYZEN`)?

Comment (Rombur) [reply cmake/kokkos_arch.cmake]: ryzen is a cpu. I don't think it has anything to do with rocm. But yeah I would remove these architectures and we can add back what we want later.

**Review by crtrott — APPROVED** [review]

**Pull request merged**
